libV2/redis.js

```diff
@@ -0,0 +1,170 @@
+    }
+
+    _initClient(startProbe = true) {
+        moduleLogger.debug('initializing redis client');
+        if (this._redis !== null) {
+            this._redis.off('connect', this._onConnect);
+            this._redis.off('ready', this._onReady);
+            this._redis.off('error', this._onError);
+        }
+        this._isConnected = false;
+        this._isReady = false;
+        this._redis = new IORedis(this._redisOptions);
+        this._redis.on('connect', this._onConnect.bind(this));
+        this._redis.on('ready', this._onReady.bind(this));
+        this._redis.on('error', this._onError.bind(this));
+        if (startProbe && this._runningRedisProbe === null) {
+            this._runningRedisProbe = setInterval(this._probeRedis.bind(this), CONNECTION_TIMEOUT);
+        }
+    }
+
+    _probeRedis() {
+        if (this.isReady) {
+            moduleLogger.debug('redis client is ready, clearing reinitialize interval');
+            clearInterval(this._runningRedisProbe);
+            this._runningRedisProbe = null;
+        } else {
+            moduleLogger.warn('redis client has failed to become ready, reinitializing');
+            this._initClient();
+        }
+    }
+
+    _onConnect() {
+        this._isConnected = true;
+        this.emit('connect');
+    }
+
+    _onReady() {
+        this._isReady = true;
+        this.emit('ready');
+    }
+
+    // eslint-disable-next-line class-methods-use-this
+    _onError(error) {
+        moduleLogger.error('error connecting to redis', { error });
+        this.emit('error', error);
+    }
+
+
+    // eslint-disable-next-line class-methods-use-this
+    _createCommandTimeout() {
+        const timerId = uuid.v4();
+        const cancel = () => {
+            if (this._inFlightTimeouts[timerId]) {
+                clearTimeout(this._inFlightTimeouts[timerId]);
+                delete this._inFlightTimeouts[timerId];
+            }
+        };
+
+        const timeout = new Promise((_, reject) => {
+            const timer = setTimeout(
+                () => {
+                    moduleLogger.warn('redis command timed out');
+                    delete this._inFlightTimeouts[timerId];
+                    this.emit('timeout');
+                    this._initClient();
+                    reject(errors.OperationTimedOut);
+                },
+                COMMAND_TIMEOUT,
+            );
+
+            this._inFlightTimeouts[timerId] = timer;
+            this.once('timeout', cancel);
+        });
+
+        return { timeout, cancel };
+    }
+
+    async _call(asyncFunc) {
+        const funcPromise = asyncFunc(this._redis);
+        if (!this._useTimeouts) {
+            // If timeouts are disabled simply return the Promise
+            return funcPromise;
+        }
+
+        const { timeout, cancel } = this._createCommandTimeout();
+
+        try {
+            const resp = await Promise.race([funcPromise, timeout]);
+            cancel();
```

Comment:
All calls to `cancel` including `this.once('timeout', cancel)` I believe might be replaced by a single unconditional `cancel()` call in a `finally {}` block, making sure we don't forget to cancel the timer in some code path (with simpler code).

Also for readability, `cancel` might be a bit confusing as it does not tell what it is canceling: `stopTimeout` or `cancelTimeout` could be better.

Reply:
I've reworked  this including to include the set changes as well, but I'm on the fence about `this.once('timeout', cancel)`. The intention is to cause concurrent requests to also timeout if any one of them does.

Reply:
Ah, isn't it better to have each request timeout independently from the other? Just wondering, otherwise if you want to keep the current behavior no problem.